### DISCUSSION TITLE
What would you recommend for installing custom binaries from `crates.io`?

### DISCUSSION THREAD

**Discussion by dpc** [ipetkov/crane#70] [Q&A] [answered]
I need to add some custom binaries from crates.io that are not in nixpkgs yet. (Basically to replace `cargo install somepackage`). I guess `crane` kind of does something similiar and there's even `downloadCargoPackage`, but I can't find a way to just build and install a crates.io package. Should I just use something entirely different?

**Comment by ipetkov** [comment, marked as answer]
You'd need to package the tool just like you would any other program. `buildPackage` can totally do that for you, the only difference being using a remote fetcher instead of a local path for `src`:

```nix
# Fetch from github:
craneLib.buildPackage rec {
  pname = "somecrate";
  version = "0.0.1";

  src = pkgs.fetchFromGitHub {
    owner = "someowner";
    repo = "somerepo";
    rev = "v${version}";
    sha256 = "...";
  };

  # set buildInputs/nativeBuildInputs as appropriate
}
```

```nix
# Fetch from crates.io
craneLib.buildPackage rec {
  pname = "somecrate";
  version = "0.0.1";

  src = pkgs.fetchCrate {
    inherit pname version;
    sha256 = "...";
  };

  # set buildInputs/nativeBuildInputs as appropriate
}
```

Then you can pass that derivation into the `buildInputs`/`nativeBuildInputs` for another derivation and it should be usable from the build script

**Reply by dpc** [reply]
Thanks a lot. Worked like a charm. I wasn't aware of `fetchCrate`. How do you discover such things?


BTW. I understand how this all work, yet it still blows my mind how easy and composable it is.

**Reply by ipetkov** [reply]
> How do you discover such things?

Not sure how I came across it, I think I was probably trying to update some Rust application in nixpkgs and found it :smile: 
fwiw the manual does mention it in the [rust specific section](https://nixos.org/manual/nixpkgs/unstable/#compiling-rust-applications-with-cargo), but as always, discoverability of the docs is a major pain point

> BTW. I understand how this all work, yet it still blows my mind how easy and composable it is.

Yeah using Nix is super awesome and this is one of the main reasons I enjoy it!

**Reply by dpc** [reply]
Thanks. I should look at the manual more often. I'm kind of aware it exist, but it is so large, it's hard to know what is there.

**Comment by marijanp** [comment]
Can we add this to the docs i.e. the FAQ?

**Reply by ipetkov** [reply]
Sure, I'd be happy to review a PR!
